### DISCUSSION TITLE
IPv6 Support in Milvus: Implementation, Enablement, and Testing Proposal

### DISCUSSION THREAD

**Discussion by codinghhr1103** [milvus-io/milvus#36191] [Q&A and General discussion] [answered]
# IPv6 Support in Milvus: Implementation, Enablement, and Testing

## I. Summary

This issue addresses the recent addition of IPv6 support to Milvus (PR #32960) and proposes methods for enabling and testing this feature. I've implemented and tested these methods, and I'm seeking feedback on their correctness and completeness.

## II. Background

1. A recent pull request (#32960) introduced IPv6 support in Milvus.
2. I've developed a method to enable IPv6 in Milvus containers.
3. I've also created a series of tests to verify IPv6 functionality.

### B. Changes:
   1. Implemented logic to handle both IPv4 and IPv6 addresses.
   2. Prioritized returning IPv4 addresses over IPv6 when available.
   3. Provided fallback to IPv6 when IPv4 is not present.

### C. Testing:
   1. Tested locally using modified versions of docker-compose files and development containers.
   2. Verified in dual-stack network setups using Docker.
   3. Checked for absence of errors in some log files.
   4. Confirmed successful connections from host machine using pymilvus, including basic operations like list_collections().
   5. Tested a version where GetValidLocalIP only returned IPv6 addresses.
   Note: Did not run all Python unit/integration tests.

### D. Impact:
   - Improves Milvus accessibility in IPv6 networks without disrupting IPv4 users.

## IV. Enabling IPv6 in Milvus

I've developed the following method to enable IPv6 in Milvus containers:

1. Locate the docker-compose.yml file.
2. Update the Milvus image version:
   image: milvusdb/milvus:v2.4.10
3. Add IPv6 support to the network configuration:
   networks:
     default:
       name: milvus
       enable_ipv6: true
4. Recreate the Milvus service:
   docker-compose down
   docker-compose up -d
5. Verify IPv6 enablement:
   docker ps -a
   docker inspect <container_id> | findstr "IPv6"

## V. Testing IPv6 Functionality

I propose the following steps to test IPv6 support in Milvus:

### A. Check host IPv6 configuration:
   ipconfig /all | findstr "IPv6"

### B. Inspect Milvus container IPv6 settings:
   1. docker inspect <container-id> | findstr "IPv6"
   2. docker port <container-id>
   3. docker network ls
   4. docker network inspect <network-name>

### C. Test inter-container IPv6 connectivity:
   1. Enter the container:
      docker exec -it <container-id> /bin/bash
   2. Install network tools:
      apt-get update && apt-get install -y iputils-ping iproute2
   3. Ping other containers:
      ping <other-container-ipv6-address>
   4. Test specific port connections:
      timeout 5 bash -c "</dev/tcp/<ipv6-address>/<port>" && echo "Port is open" || echo "Port is closed"

### D. Test database operations over IPv6:
   1. Disable IPv4 forwarding in the container:
      sysctls:
        - net.ipv4.ip_forward=0
   2. Verify IPv4 forwarding is disabled:
      sysctl net.ipv4.ip_forward
      cat /proc/sys/net/ipv4/ip_forward
   3. Run Python scripts to perform Milvus operations (slicing, embedding, storage, similarity calculation, query results) over IPv6.

## VI. Questions and Requests

1. Are the methods I've outlined for enabling IPv6 in Milvus containers correct and optimal?
2. Do the proposed testing procedures adequately cover IPv6 functionality?
3. Are there any additional tests or configurations that should be considered?

Your feedback and suggestions would be greatly appreciated. Thank you for your time and attention to this matter.

**Comment by yhmo** [comment]
The pr #32960 has been merged into master branch but not cherry-picked to 2.4 branch.

Wait @xiaofan-luan comment.

**Comment by pingliu** [comment]
Currently, the restful api supports ipv6, but the sdk does not support it yet.
the version of milvus tested is v2.4.11.

you can test like this:
```
export MILVUS_URI="http://[ipv6_addr]:19530"
curl  --request GET \
    --url "${MILVUS_URI}/v1/vector/collections" \
    --header 'accept: application/json' \
    --header 'content-type: application/json'
```

more restful api: https://milvus.io/api-reference/restful/v2.4.x/About.md

**Reply by codinghhr1103** [reply]
When I was using the PyMilvus SDK, using IPV6 caused some issues. Thanks for pointing out the Restful API, which I tested and does support ipv6-based connections.

**Comment by yhmo** [comment]
@codinghhr1103 
Did you ever cherry-pick the pr #32960 to the 2.4 branch? I didn't see it in the 2.4 branch.

**Comment by codinghhr1103** [comment]
As shown in the two attached images, the changes introduced in PR #32960 are currently available only in the master branch but have not yet been incorporated into the 2.4 release. To leverage this feature, I am compiling the master branch from source and creating a Docker image for deployment via Docker Compose. 

![1](https://github.com/user-attachments/assets/7efd1990-1c87-4f2d-910f-aa41be7d0758)

![2](https://github.com/user-attachments/assets/85dc7857-3bd1-471c-b5b5-64b2d9131c85)

Additionally, when attempting to connect to the database using the PyMilvus SDK, I've encountered an issue with passing an IPv6 address due to the presence of colons in the address format. I plan to test connecting directly to the database using the Restful API with an IPv6 address to determine if this resolves the issue.

**Reply by yhmo** [reply, marked as answer]
If the pr is only in the master branch, I suppose this feature is available in v2.5.x.

You don't have to build the image by yourself.
There are new images built from the master branch for each couple of days: https://hub.docker.com/r/milvusdb/milvus/tags
The latest image is:
`docker pull milvusdb/milvus:master-20240914-ef8289ea-amd64`

**Reply by codinghhr1103** [reply]
I'm looking forward to the 2.5 release.

Thanks for pointing out the image address.
